docs(feat): Zqs 1060 write backends guide (#33)
zapatacomputing/orquestra-core: docs/guides/backends.rst

.. _backends_guide:

==============
Backends guide
==============

> Backends are the interfaces used to run :ref:`Circuits <circuits_guide>` in Orquestra Core. In this guide we'll cover differences between types of backends (``QuantumBackend`` vs ``QuantumSimulator``, ``TrackingBackend``, ``SymbolicSimulator``), how to use backends, currently integrated backends, and creating your own backend (among other topics).

Reviewer: I would drop `TrackingBackend` and `SymbolicSimulator` from here.

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64

General Backend Information
===========================

The function of a backend is to send instructions to quantum hardware or to a simulator. Since there are various types of quantum hardware and simulators, different sets of instructions needs to be written and it can become complex. ``QuantumBackend`` simplifies this process by enabling the user to deploy a single set of instructions on multiple hardware and simulators simultaneously. Therefore executing code on various devices becomes easier and more manageable.

Two types of backends
---------------------

> ``QuantumBackend`` is the base class for all backends. ``QuantumBackend``\ s are used for physical quantum hardware backends.

Reviewer: Well, that's not technically correct, as `QuantumSimulator` inherits from `QuantumBackend`.

```suggestion
``QuantumBackend`` is the base class for all backends. ``QuantumBackend``\ s can be used for physical quantum hardware backends as well as simulators.
```

Because ``QuantumBackend``\ s connect to a cloud provider to run remote circuits, it is necessary to specify which device you want to connect to. For example, the ``QiskitBackend`` takes in a ``device_name`` string:

.. literalinclude:: /examples/backends_guide.py
  :start-after: # QuantumBackend creation example
  :end-before: # End QuantumBackend creation example

.. _quantum_simulator:

``QuantumSimulator`` is a subclass of ``QuantumBackend`` with added functionality that only simulators can provide, such as getting the wavefunction of a circuit and getting exact expectation values. Unlike ``QuantumBackend``\ s, ``QuantumSimulator``\ s usually do not require a ``device_name`` parameter. One exception to this rule is the ``QiskitSimulator``, which has :ref:`multiple options <integrated_backends>`:

.. literalinclude:: /examples/backends_guide.py
  :start-after: # QuantumSimulator creation example
  :end-before: # End QuantumSimulator creation example

For more information on how to use each type of backend please refer to the :ref:`section in this guide on various methods <backend_methods>` or to the :doc:`API documentation </api/quantum/api/backend/index>`.

> Testing the Backends

Reviewer: I would move this section at the bottom, perhaps before "How to integrate your own backend".

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64 moved to be a subsection of integrating backends. I couldn't think of a (common) scenario other than integrating a new backend where this would be helpful info, feel free to correct me ;)

--------------------

> The `backend tests <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/api/backend_test.py>`_ cover both general behaviors we want to make sure the ``Backend``\ s follow and specific gate tests. For instance, a common "general behavior" test across multiple backends is the ``run_circuitset_and_measure`` test. Here are examples in the `Cirq tests <https://github.com/zapatacomputing/orquestra-cirq/blob/main/tests/orquestra/integrations/cirq/simulator/simulator_test.py#L52=>`_ and the `Qiskit tests <https://github.com/zapatacomputing/orquestra-qiskit/blob/main/tests/orquestra/integrations/qiskit/backend/backend_test.py#L162=>`_.

Reviewer: I'd add an intro paragraph here, sth along these lines:

"One of our goals with the QuantumBackends was to make sure that all the backends behave in a uniform way. In order to ensure that, we have introduced a suite of tests which all backends need to pass."

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64 and also moved testing the backends to be a subsection of integrating the backends (I think that makes sense!)

.. note::

>     Because tests are often specific to their ``Backend``, there are only prototype backend tests in `orquestra-quantum <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/api/backend_test.py>`_. Implementations of these tests are in the integrations that house specific backends. Please refer to the section on :ref:`currently integrated backends <integrated_backends>` for a complete list of backends

Reviewer: I don't think the old statement was correct. 
```suggestion
    Because tests are often specific to their ``Backend``, there are only basic tests in `orquestra-quantum <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/api/backend_test.py>`_, which might be extended for specific implementations.
```

> For gates tests, the goal is to ensure gates are properly implemented. Sometimes gate operations in certain simulators are in reverse order (compared to Orquestra Core and other frameworks). When you use a gate operation in Orquestra, it's important to make sure it's represented correctly in the simulator and/or backend. In order to do that, Orquestra Core has a `suite of gate operations <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/testing/test_cases_for_backend_tests.py>`_ that should be tested on each backend and simulator. This is done with `QuantumBackendGateTests <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/api/backend_test.py#L193=>`_ and `QuantumSimulatorGateTests <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/api/backend_test.py#L432=>`_

Reviewer: ```suggestion
For gates tests, the goal is to ensure gates are properly implemented and always follow the same conventions. Sometimes gate operations in certain simulators are in reverse order (compared to Orquestra Core and other frameworks). When you use a gate operation in Orquestra, it's important to make sure it's represented correctly in the simulator and/or backend. In order to do that, Orquestra Core has a `suite of gate operations <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/testing/test_cases_for_backend_tests.py>`_ that should be tested on each backend and simulator. This is done with `QuantumBackendGateTests <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/api/backend_test.py#L193=>`_ and `QuantumSimulatorGateTests <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/api/backend_test.py#L432=>`_
```

.. code-block:: python

  class TestCirqSimulatorGates(QuantumSimulatorGatesTest):
      gates_to_exclude = ["RH"]
      pass

.. _backend_methods:

Backend Methods Usage
=====================

In order to use quantum hardware, the user must use the backend's authentication method to prove that they have the right credentials to send instruction to the hardware. The authentication method is specified by the hardware provider. Providing authentication credentials is the first step in using ``QuantumBackend``. Below is an example of using IBM's hardware through ``QiskitBackend``.

``QuantumSimulator``\ s run in your own environment and therefore can be used without needing authentication credentials.

.. literalinclude:: /examples/backends_guide.py
  :start-after: # QuantumBackend creation example
  :end-before: # End QuantumBackend creation example

General Backend Methods
-----------------------

> After intializing the backend we can send the ``Circuit`` and number of repetitions to the backend and read the measurement. For more information on creating circuits to send to a backend, refer to the :ref:`Circuits Guide <circuits_guide>`. For a single circuit, that is executed as follows:

Reviewer: Suggestion?
```suggestion
After intializing the backend we can send the ``Circuit`` and number of repetitions to the backend and read the measurement. For more information on creating circuits to send to a backend, refer to the :ref:`Circuits Guide <circuits_guide>`. For a single circuit, it looks as follows:
```

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64 with slightly different language :)

.. literalinclude:: /examples/backends_guide.py
  :start-after: # QuantumBackend run and measure circuit
  :end-before: # End QuantumBackend run and measure circuit

If you want to run multiple circuits with different number of samples, the code needs to be modified as follows:

.. literalinclude:: /examples/backends_guide.py
  :start-after: # QuantumBackend run and measure circuitset
  :end-before: # End QuantumBackend run and measure circuitset

``QuantumBackend`` can also provide the distribution of the outcome measurement. It can be extracted using the following approach:

.. literalinclude:: /examples/backends_guide.py
  :start-after: # Quantumbackend measurement distribution
  :end-before: # End Quantumbackend measurement distribution

In all instances, ``QuantumBackend`` automatically adds the measurement operators to the circuit. Therefore, the user is not required to include it in the circuit.

> Simulator-specific methods

Reviewer: I think we're lacking information about using simulations with noise?
It would be good to mention it somewhere, perhaps in this section?

Sth like:
"Some simulators, such as Qiskit or QSim allow for running simulations with various noise models – for more details how to use that, please refer to docs for particular simulator".

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64

--------------------------

> ``Quantumsimulator`` can be used to extract the wave function and expectation value of a given circuit. Below is an example of extracting these values using a ``Quantumsimulator`` called ``CirqSimulator``:

Reviewer: Expectation value is not a property of a circuit. Perhaps sth like this?
```suggestion
``Quantumsimulator`` can be used to extract a wave function and expectation value for a given circuit (or circuit and operator). Below is an example of extracting these values using a ``Quantumsimulator`` called ``CirqSimulator``:
```

.. literalinclude:: /examples/backends_guide.py
  :start-after: # Quantumsimulator examples
  :end-before: # End Quantumsimulator examples

Zapata's backends
=================

SymbolicSimulator
-----------------

> Zapata's ``SymbolicSimulator`` is a type of :ref:`quantum simulator <quantum_simulator>` that's built specifically to allow for the evaluation of circuits with :ref:`symbolic gates <symbolic_gates>`. Because there is extra overhead to allow for this broader type of evaluation, ``SymbolicSimulator`` is not the most performant simulator, but it can be useful for deeply understanding small circuits.

Reviewer: I'd also add sth along the lines:

As it doesn't require any extra dependencies apart from `orquestra-quantum`, it can be also used for testing purposes.

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64

The ``SymbolicSimulator`` can be used the same way any other simulator can be. To see more usage please refer to the :ref:`section in this guide on various methods <backend_methods>` or to the :doc:`API documentation </api/quantum/api/backend/index>`

> TrackingBackend

Reviewer: I don't think `TrackingBackend` is a broader class of backends.
I would just drop the whole `TrackingBackend` thing from this convo and just leave it to be about `MeasurementTrackingBackend`.

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64

---------------

A ``TrackingBackend`` is a backend that tracks and stores data about each run of a circuit. This is accomplished by wrapping pre-existing backends in a ``TrackingBackend``. Currently the ``MeasurementTrackingBackend`` in :doc:`orquestra.quantum.trackers </api/quantum/trackers/index>` is the only implemented ``TrackingBackend``

> When ``run_circuit_and_measure`` is called, the tracking backend stores the ``data_type`` recieved from the call (usually a measurement outcome distribution), the wrapped device, the circuit which was run, the disribution of bitstrings which was recieved, the number of gates in the circuit, and the number of shots run in a JSON file. This is useful when you are paying to run circuits on a machine and you want to be able to re-use the data in a different workflow.

Reviewer: ```suggestion
When ``run_circuit_and_measure`` is called, the tracking backend stores the ``data_type`` received from the call (usually a measurement outcome distribution), the wrapped device, the circuit which was run, the distribution of bitstrings which was received, the number of gates in the circuit, and the number of shots run in a JSON file. This is useful when you are paying to run circuits on a machine and you want to be able to re-use the data for different computations.
```

> To create a ``MeasurementTrackingBackend`` it must be initialized with a backend to be wrapped around, a name for the file you are storing the data in, and a boolean indicating whether or not the individual bitstrings should be saved. 

Reviewer: ```suggestion
``MeasurementTrackingBackend`` must be initialized with a backend to be wrapped around, a name for the file you are storing the data in, and a boolean indicating whether or not the individual bitstrings should be saved. 
```

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64 with slightly different language

> .. literalinclude:: /examples/backends_guide.py

Reviewer: Since we're talking about the boolean in the description above, it should be include in the example as well.

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64 with different language (because the boolean is optional)

  :start-after: # TrackingBackend creation example
  :end-before: # End TrackingBackend creation example

.. caution:: 
  
  If all the bitstrings are saved, it will greatly increase the size of the produced JSON file.

To use the ``MeasurementTrackingBackend`` you can call ``run_circuit_and_measure`` or ``run_circuitset_and_measure`` :ref:`just like with all other QuantumBackends <backend_methods>`. There is also the option to manually append measurement information to the data in a ``MeasurementTrackingBackend`` using ``record_raw_measurement_data``. This is not usually necessary, but if your use case requires it, please see the :doc:`API documentation </api/quantum/trackers/index>` for more information. 

For more examples of using the ``MeasurementTrackingBackend``, refer to the `tracking backend tests <https://github.com/zapatacomputing/orquestra-quantum/blob/main/tests/orquestra/quantum/trackers_test.py>`_

.. _integrated_backends:

Currently integrated backends
=============================

> Hardware backends

Reviewer: I would add a little bit more info about these backends, so it's easier for people to see what they should be using?

Reviewer: Let's just leave it for now, we'll add more details on the next iteration.

There are slightly different ways to import the conversions from the integrations for different frameworks, so here's examples for all of them:

.. literalinclude:: /examples/backends_guide.py
  :start-after: # Importng and Exporting with different frameworks
  :end-before: # end importing/exporting examples

If you want to use some specific features from a particular framework (e.g. drawing circuits from qiskit) feel free to export/import and do it! Examples of using the import and export functions for this purpose can be found in the :ref:`getting started tutorial <beginner_translating_circuits>`.

How to integrate your own backend
=================================

We have simplified the process of integrating any simulator into Orquestra Core. First you need to create a function that can translate gates between your library and `Orquestra's gates <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/circuits/_builtin_gates.py>`_. Then create a class method that inherits from ``QuantumSimulator`` and overrides the needed abstract methods (like ``run_circuit_and_measure``) using the approach below:

> 

Reviewer: I'd add sth like:
"In order for integration to work properly, the signature of all the methods needs to be exactly the same as in the abstract class. If your simulator requires some extra arguments (e.g. whether to use a noise model), please provide it in the `__init__` method."

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64

> .. literalinclude:: /examples/backends_guide.py

Reviewer: I think the argument names in the sample code are not correct. I don't think it's `shots`.

Author: fixed in 4551c621e5c9c2b2295716c545eb6211b836df64
